File: project_euler/problem_903/solution.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * Problem 903: Matrix Permanent Modulo a Prime
 *
 * A[i][j] = (i*j) mod n for i,j in {1,...,n}, n = 12.
 * Find perm(A) mod 10^9+7.
 *
 * Key insight: Row n has A[n][j] = (n*j) mod n = 0 for all j.
 * A matrix with a zero row has permanent = 0.
 *
 * Two methods:
 *   1. Zero-row detection (O(n) check)
 *   2. Ryser's formula (O(2^n * n) general computation)
 */

const long long MOD = 1e9 + 7;

/*
 * Method 1: Check for zero row
 * If any row i has A[i][j] = 0 for all j, perm = 0.
 * For the multiplication table mod n, row n always satisfies this.
 */
bool has_zero_row(int n) {
    for (int i = 1; i <= n; i++) {
        bool all_zero = true;
        for (int j = 1; j <= n; j++) {
            if ((i * j) % n != 0) {
                all_zero = false;
                break;
            }
        }
        if (all_zero) return true;
    }
    return false;
}

/*
 * Method 2: Ryser's formula
 * perm(A) = (-1)^n * sum_{S} (-1)^|S| * prod_i (sum_{j in S} A[i][j])
 *
 * O(2^n * n) time.  For n=12, 2^12 * 12 = 49152 operations.
 */
long long permanent_ryser(vector<vector<int>>& A) {
    int n = A.size();
    long long result = 0;

    for (int mask = 1; mask < (1 << n); mask++) {
        int bits = __builtin_popcount(mask);

        // Product of row sums for columns in mask
        long long prod = 1;
        bool zero = false;
        for (int i = 0; i < n && !zero; i++) {
            long long row_sum = 0;
            for (int j = 0; j < n; j++) {
                if (mask & (1 << j))
                    row_sum += A[i][j];
            }
            if (row_sum == 0) { zero = true; break; }
            prod *= row_sum;
        }
        if (zero) continue;

        long long sign = ((n - bits) % 2 == 0) ? 1 : -1;
        result += sign * prod;
    }

    // Multiply by (-1)^n
    if (n % 2 == 1) result = -result;
    return result;
}

int main() {
    int n = 12;

    // Method 1: structural
    if (has_zero_row(n)) {
        cout << 0 << endl;
    }

    // Method 2: Ryser verification
    vector<vector<int>> A(n, vector<int>(n));
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            A[i][j] = ((i + 1) * (j + 1)) % n;

    long long perm = permanent_ryser(A);
    assert(perm == 0);

    // Verify for small n: all should be 0
    for (int nn = 2; nn <= 8; nn++) {
        vector<vector<int>> M(nn, vector<int>(nn));
        for (int i = 0; i < nn; i++)
            for (int j = 0; j < nn; j++)
                M[i][j] = ((i + 1) * (j + 1)) % nn;
        assert(permanent_ryser(M) == 0);
    }

    cout << 0 << endl;
    return 0;
}

File: project_euler/problem_903/solution.py
"""
Problem 903: Matrix Permanent Modulo a Prime

Let A be the n x n matrix with A[i,j] = (i*j) mod n, i,j in {1,...,n}.
Find perm(A) mod 10^9+7 for n = 12.

Key insight: Row n has A[n,j] = (n*j) mod n = 0 for all j, so perm(A) = 0.

Methods:
    1. Structural argument (zero row detection)
    2. Ryser's formula (general permanent computation)
    3. Brute-force via itertools.permutations (small n verification)
"""

from itertools import permutations
from math import gcd

MOD = 10**9 + 7

def solve_structural(n: int):
    """Detect if any row is all zeros => perm = 0."""
    for i in range(1, n + 1):
        if all((i * j) % n == 0 for j in range(1, n + 1)):
            return 0  # Row i is all zeros
    return None  # No zero row found; need computation

def permanent_ryser(A: list, mod: int = 0):
    """Compute permanent using Ryser's inclusion-exclusion formula.

    perm(A) = (-1)^n * sum_{S subset [n]} (-1)^|S| * prod_{i} sum_{j in S} A[i][j]
    Complexity: O(2^n * n).
    """
    n = len(A)
    result = 0
    for mask in range(1, 1 << n):
        bits = bin(mask).count("1")
        # Compute product of row sums restricted to columns in mask
        prod = 1
        for i in range(n):
            row_sum = 0
            for j in range(n):
                if mask & (1 << j):
                    row_sum += A[i][j]
            prod *= row_sum
            if prod == 0:
                break  # Early termination
        sign = (-1) ** (n - bits)
        result += sign * prod
    result *= (-1) ** n
    return result % mod if mod else result

def permanent_brute(A: list):
    """Compute permanent by definition: sum over all permutations."""
    n = len(A)
    total = 0
    for sigma in permutations(range(n)):
        prod = 1
        for i in range(n):
            prod *= A[i][sigma[i]]
            if prod == 0:
                break
        total += prod
    return total

# Build matrix and solve
N = 12
A = [[(i * j) % N for j in range(1, N + 1)] for i in range(1, N + 1)]

ans_structural = solve_structural(N)
assert ans_structural == 0, "Structural method failed"

ans_ryser = permanent_ryser(A) % MOD

for test_n in [3, 4, 5, 6, 7]:
    M = [[(i * j) % test_n for j in range(1, test_n + 1)] for i in range(1, test_n + 1)]
    p_ryser = permanent_ryser(M)
    p_brute = permanent_brute(M)
    assert p_ryser == p_brute, f"n={test_n}: Ryser={p_ryser}, brute={p_brute}"
    assert p_ryser == 0, f"n={test_n}: permanent should be 0, got {p_ryser}"

assert ans_ryser == 0

print(0)

# Permanents of the reduced matrix (i,j in {1..n-1}) for prime n
def permanent_reduced(n: int):
    """Permanent of the (n-1)x(n-1) matrix B[i,j] = (i*j) mod n, i,j in {1..n-1}."""
    B = [[(i * j) % n for j in range(1, n)] for i in range(1, n)]
    return permanent_ryser(B)
